metrics/src/common.rs

```diff
@@ -0,0 +1,235 @@
+                parts.join(".")
+            }
+        }
+    }
+}
+
+pub(crate) type MetricScopeHandle = u64;
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub(crate) enum MetricKind {
+    Counter,
+    Gauge,
+    Histogram,
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub(crate) enum MetricIdentifier {
+    Unlabeled(MetricName, MetricScopeHandle, MetricKind),
+}
+
+#[derive(Debug)]
+enum ValueState {
+    Counter(AtomicU64),
+    Gauge(AtomicI64),
+    Histogram(AtomicWindowedHistogram),
+}
+
+#[derive(Debug)]
+pub(crate) enum ValueSnapshot {
+    Counter(u64),
+    Gauge(i64),
+    Histogram(StreamingIntegers),
+}
+
+#[derive(Clone, Debug)]
+/// Handle to the underlying measurement for a metric.
+pub(crate) struct MetricValue {
+    state: Arc<ValueState>,
+}
+
+impl MetricValue {
+    fn new(state: ValueState) -> Self {
+        MetricValue {
+            state: Arc::new(state),
+        }
+    }
+
+    pub(crate) fn new_counter() -> Self {
+        Self::new(ValueState::Counter(AtomicU64::new(0)))
+    }
+
+    pub(crate) fn new_gauge() -> Self {
+        Self::new(ValueState::Gauge(AtomicI64::new(0)))
+    }
+
+    pub(crate) fn new_histogram(window: Duration, granularity: Duration, clock: Clock) -> Self {
+        Self::new(ValueState::Histogram(AtomicWindowedHistogram::new(
+            window,
+            granularity,
+            clock,
+        )))
+    }
+
+    pub(crate) fn update_counter(&self, value: u64) {
+        match self.state.deref() {
+            ValueState::Counter(inner) => {
+                inner.fetch_add(value, Ordering::Release);
+            }
+            _ => unreachable!("tried to access as counter, not a counter"),
+        }
+    }
+
+    pub(crate) fn update_gauge(&self, value: i64) {
+        match self.state.deref() {
+            ValueState::Gauge(inner) => inner.store(value, Ordering::Release),
+            _ => unreachable!("tried to access as gauge, not a gauge"),
+        }
+    }
+
+    pub(crate) fn update_histogram(&self, value: u64) {
+        match self.state.deref() {
+            ValueState::Histogram(inner) => inner.record(value),
+            _ => unreachable!("tried to access as histogram, not a histogram"),
+        }
+    }
+
+    pub(crate) fn snapshot(&self) -> ValueSnapshot {
+        match self.state.deref() {
+            ValueState::Counter(inner) => {
+                let value = inner.load(Ordering::Acquire);
+                ValueSnapshot::Counter(value)
+            }
+            ValueState::Gauge(inner) => {
+                let value = inner.load(Ordering::Acquire);
+                ValueSnapshot::Gauge(value)
+            }
+            ValueState::Histogram(inner) => {
+                let stream = inner.snapshot();
+                ValueSnapshot::Histogram(stream)
+            }
+        }
+    }
+}
+
+/// Trait for types that represent time and can be subtracted from each other to generate a delta.
+pub trait Delta {
+    fn delta(&self, other: Self) -> u64;
```

Comment:
we should add `warn` docs cuz this needs a doc comment 😄

Reply:
I added `#![deny(missing_docs)]` to really make sure we don't slip up... and I added a doc comment for `delta`. :)